### PR TITLE
Use shared CI tooling installers

### PR DESCRIPTION
**Pull request by leynos**
## Summary

This branch replaces handwritten CI provisioning of Nixie and Whitaker with
maintained `leynos/shared-actions` installers, while preserving Whitaker
installer version `0.2.7`. It bootstraps Rust `1.95.0` only while Nixie installs
Merman CLI, then restores the project's supported Rust `1.85.0` toolchain
before Whitaker and repository gates run.

It pins every shared-actions consumer to
`794e4801babcf68065c660fdf4781ad62be5d061`, the latest verified upstream SHA,
so all reusable workflows and composite actions resolve the same revision.

## Review walkthrough

- Start with [the lint job installer setup](https://github.com/leynos/cuprum/blob/use-shared-tooling-installers/.github/workflows/ci.yml#L37), which bootstraps Nixie with Rust `1.95.0` and restores Rust `1.85.0` for project gates.
- Then review [the main coverage action pins](https://github.com/leynos/cuprum/blob/use-shared-tooling-installers/.github/workflows/coverage-main.yml#L49), [Dependabot automation](https://github.com/leynos/cuprum/blob/use-shared-tooling-installers/.github/workflows/dependabot-automerge.yml#L31), and [mutation testing](https://github.com/leynos/cuprum/blob/use-shared-tooling-installers/.github/workflows/mutation-testing.yml#L29) for the consistent SHA update.
- Finish with [the CI contract test](https://github.com/leynos/cuprum/blob/use-shared-tooling-installers/cuprum/unittests/test_extension_ci_contract.py#L168), which protects the action pins, toolchain versions, and installation order.

## Validation

- `make check-fmt`: passed
- `make lint`: passed
- `make typecheck`: passed
- `make test`: passed
- `make markdownlint`: passed
- `make nixie`: passed
- `git diff --check origin/main...HEAD`: passed

## Summary by Sourcery

Adopt maintained shared-actions installers throughout CI while preserving toolchain compatibility and enforcing the resulting workflow contract.

Enhancements:
- Replace handwritten Nixie and Whitaker provisioning in CI with shared maintained installers while preserving the supported Rust toolchain for project gates.
- Add CI contract coverage for installer ordering, toolchain versions, cache behavior, and immutable shared-action references.
- Defer workflow helper imports during isolated test collection to keep script tests independent of workflow-parser dependencies.

CI:
- Standardize shared-actions usage across CI, coverage, Dependabot automerge, and mutation-testing workflows with immutable commit references.

Documentation:
- Update developer guidance to document shared installer usage, temporary Rust bootstrapping, and immutable action-reference testing.

Tests:
- Extend workflow and cache tests to cover the additional Nixie toolchain setup and shared installer requirements.

Enhancements:

- Replace handwritten Nixie and Whitaker CI provisioning with maintained shared-actions installers while preserving the project’s supported Rust toolchain for repository gates.
- Standardize shared-actions consumers on a single verified commit and retain the required Rust toolchain and installer ordering across CI workflows.
- Extend CI contract coverage to enforce shared-action pins, toolchain versions, installation order, and cache configuration.

CI:

- Update coverage, Dependabot automerge, mutation testing, and general CI workflows to use the verified shared-actions revision.

Tests:

- Update CI cache and workflow tests for the additional Nixie toolchain setup and shared installer requirements.

## References

- [Lody session](https://lody.ai/leynos/sessions/1693aaba-db04-470d-830f-3eedd5db1f6d)

### CODE REVIEW

**Comment by buzzybee-df12** [comment]
@coderabbitai review

**Comment by buzzybee-df12** [comment]
@coderabbitai review

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -148,39 +148,29 @@ jobs:
...
-          toolchain: '1.85.0'
+          toolchain: '1.95.0'
           cache-provider: external
           use-sccache: 'false'
 
-      - name: Cache Whitaker installation
-        id: cache-whitaker
-        uses: actions/cache@55cc8345863c7cc4c66a329aec7e433d2d1c52a9 # v6.1.0
+      - name: Install Nixie
+        uses: leynos/shared-actions/.github/actions/install-nixie@794e4801babcf68065c660fdf4781ad62be5d061
```
[the remark this answers is not in the document]
Comment (leynos) [reply .github/workflows/ci.yml]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response. Ensure that this is validated against the *current* version of the codegraph.

If further refinement to address this finding would be deleterious, please supply a clear explanatory one to two paragraph markdown message in a code block that I can paste into the CodeScene web ui's diagnostic suppression function so this diagnostic can be silenced.

[cuprum/unittests/test_extension_ci_contract.py](https://github.com/leynos/cuprum/pull/325/files/0148769ed94666bfa4f71ee6bb2c94072aa3fc34#diff-beb749b2fe2e20cdc84b46c41c76f4facbb74fc34435e15ade6cbab939ba180d)

Comment on lines +325 to +376

~~~python
def test_workflows_pin_shared_actions_to_one_revision() -> None:
    """Every shared-actions caller uses one revision without fixing its value."""
    all_uses_by_workflow = {
        workflow_name: _shared_action_uses(workflow_document(workflow_name))
        for workflow_name, _ in workflow_sources()
    }
    uses_by_workflow = {
        workflow_name: uses
        for workflow_name, uses in all_uses_by_workflow.items()
        if uses
    }
    parsed_references = {
        workflow_name: [
            (uses, SHARED_ACTIONS_REFERENCE.fullmatch(uses)) for uses in uses_values
        ]
        for workflow_name, uses_values in uses_by_workflow.items()
    }
    invalid_references = [
        f"{workflow_name}:{uses}"
        for workflow_name, references in parsed_references.items()
        for uses, reference in references
        if reference is None
    ]
    assert not invalid_references, (
        "shared-actions references must include a non-whitespace, non-comment "
        f"revision after '@': {invalid_references}"
    )
    valid_references = {
        workflow_name: [
            reference for _, reference in references if reference is not None
        ]
        for workflow_name, references in parsed_references.items()
    }

    targets_by_workflow = {
        workflow_name: {reference["target"] for reference in references}
        for workflow_name, references in valid_references.items()
    }
    assert targets_by_workflow == EXPECTED_SHARED_ACTIONS_TARGETS, (
        "each workflow must call only its expected shared actions or reusable "
        f"workflows; found {targets_by_workflow}"
    )

    revisions = {
        reference["reference"]
        for references in valid_references.values()
        for reference in references
    }
    assert len(revisions) == 1, (
        "every shared-actions call must use the same revision; "
        f"found {sorted(revisions)}"
    )
~~~

❌ New issue: [Complex Method](https://codescene.io/projects/74471/delta?repo-id=1354509&review-id=325&biomarker=Complex+Method&filename=cuprum%2Funittests%2Ftest_extension_ci_contract.py&method=test_workflows_pin_shared_actions_to_one_revision)
test_workflows_pin_shared_actions_to_one_revision has a cyclomatic complexity of 13, threshold = 9

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response. Ensure that this is validated against the *current* version of the codegraph.

If further refinement to address this finding would be deleterious, please supply a clear explanatory one to two paragraph markdown message in a code block that I can paste into the CodeScene web ui's diagnostic suppression function so this diagnostic can be silenced.

[cuprum/unittests/test_extension_ci_contract.py](https://github.com/leynos/cuprum/pull/325/files/ac03f65ae9aa71f83d84e0e193520522df02c0a4#diff-beb749b2fe2e20cdc84b46c41c76f4facbb74fc34435e15ade6cbab939ba180d)

Comment on lines +349 to +390

~~~python
def test_workflows_pin_shared_actions_to_immutable_revisions() -> None:
    """Every shared-actions caller uses an immutable revision without fixing it."""
    all_uses_by_workflow = {
        workflow_name: _shared_action_uses(workflow_document(workflow_name))
        for workflow_name, _ in workflow_sources()
    }
    uses_by_workflow = {
        workflow_name: uses
        for workflow_name, uses in all_uses_by_workflow.items()
        if uses
    }
    parsed_references = {
        workflow_name: [
            (uses, SHARED_ACTIONS_REFERENCE.fullmatch(uses)) for uses in uses_values
        ]
        for workflow_name, uses_values in uses_by_workflow.items()
    }
    invalid_references = [
        f"{workflow_name}:{uses}"
        for workflow_name, references in parsed_references.items()
        for uses, reference in references
        if reference is None
    ]
    assert not invalid_references, (
        "shared-actions references must use a full lowercase 40-character "
        f"commit SHA after '@': {invalid_references}"
    )
    valid_references = {
        workflow_name: [
            reference for _, reference in references if reference is not None
        ]
        for workflow_name, references in parsed_references.items()
    }

    targets_by_workflow = {
        workflow_name: {reference["target"] for reference in references}
        for workflow_name, references in valid_references.items()
    }
    assert targets_by_workflow == EXPECTED_SHARED_ACTIONS_TARGETS, (
        "each workflow must call only its expected shared actions or reusable "
        f"workflows; found {targets_by_workflow}"
    )
~~~

❌ New issue: [Complex Method](https://codescene.io/projects/74471/delta?repo-id=1354509&review-id=325&biomarker=Complex+Method&filename=cuprum%2Funittests%2Ftest_extension_ci_contract.py&method=test_workflows_pin_shared_actions_to_immutable_revisions)
test_workflows_pin_shared_actions_to_immutable_revisions has a cyclomatic complexity of 11, threshold = 9

**Comment by buzzybee-df12** [comment]
@coderabbitai review

**Comment by leynos** [comment]
@coderabbitai re:

### ❌ Failed check (1 error)

| Check name | Status | Explanation | Resolution |
| :---: | :--- | :--- | :--- |
| Testing (Unit And Behavioural) | ❌ Error | The added tests cover the YAML contract, regex edge cases, ordering, toolchain inputs, and cache invariants. They do not exercise the changed workflow at its functional boundary. The pull request repl… | Add a black-box workflow integration test for the changed `lint-test` path. Run the workflow with a controlled event, assert a successful exit, and verify observable outputs such as the installed tool versions, available commands, and relev… |

Don't request tests for upstream composite actions. These are already tested in `leynos/shared-actions`.

**Comment by leynos** [comment]
@coderabbitai Please investigate the cause of the following issue using codegraph exploration and research, identify a fix and provide an AI coding agent prompt for the fix:

~~~
Run make test-markdown-format
  make test-markdown-format
  shell: /usr/bin/bash -e {0}
  env:
    RUFF_VERSION: 0.16.4
    TY_VERSION: 0.0.74
    MARKDOWNLINT_VERSION: 0.20.0
    LINUX_RUNNER_VCPUS: 2
    CACHE_GENERATION: v1
    UBUNTU_RELEASE: ubuntu2404
    MBAKE_VERSION: 1.4.6
    MDTABLEFIX_VERSION: 0.5.1
    WHITAKER_INSTALLER_VERSION: 0.2.7
    UV_CACHE_DIR: /home/runner/work/cuprum/cuprum/.uv-cache
    UV_TOOL_DIR: /home/runner/work/cuprum/cuprum/.uv-tools
    UV_TOOL_BIN_DIR: /home/runner/work/cuprum/cuprum/.uv-bin
    YAMLLINT_VERSION: 1.38.0
    CARGO_CACHE_PREFIX: cargo-v1-Linux-X64-github-hosted-
    CARGO_CACHE_KEY: cargo-v1-Linux-X64-github-hosted-529e94524d6bd3f6fd6bf36247ffd9d668f08de048eb5734748bfdbb63ff0897
    TOOL_CACHE_PREFIX: tool-v1-Linux-X64-github-hosted-ubuntu2404-py3.13-
    TOOL_CACHE_KEY: tool-v1-Linux-X64-github-hosted-ubuntu2404-py3.13-28c937537e530f6bce89c2f354b03b8f737d45d2777071c7625340135a464e02
    SCCACHE_CACHE_PREFIX: sccache-v1-Linux-X64-github-hosted-ubuntu2404-py3.13-lint-
    SCCACHE_CACHE_KEY: sccache-v1-Linux-X64-github-hosted-ubuntu2404-py3.13-lint-33998594606
    pythonLocation: /opt/hostedtoolcache/Python/3.13.15/x64
    PKG_CONFIG_PATH: /opt/hostedtoolcache/Python/3.13.15/x64/lib/pkgconfig
    Python_ROOT_DIR: /opt/hostedtoolcache/Python/3.13.15/x64
    Python2_ROOT_DIR: /opt/hostedtoolcache/Python/3.13.15/x64
    Python3_ROOT_DIR: /opt/hostedtoolcache/Python/3.13.15/x64
    LD_LIBRARY_PATH: /opt/hostedtoolcache/Python/3.13.15/x64/lib
    UV_PYTHON_INSTALL_DIR: /home/runner/work/_temp/uv-python-dir
    RUSTC_WRAPPER: /home/runner/.local/bin/sccache
    SCCACHE_DIR: /home/runner/.cache/sccache
    SCCACHE_CACHE_SIZE: 4G
    CARGO_INCREMENTAL: 0
    CARGO_PROFILE_DEV_DEBUG: 0
    CARGO_TERM_COLOR: always
    RUST_BACKTRACE: short
    RUSTFLAGS: -D warnings
    CARGO_UNSTABLE_SPARSE_REGISTRY: true
    CARGO_REGISTRIES_CRATES_IO_PROTOCOL: sparse
Installed 7 packages in 7ms
ImportError while loading conftest '/home/runner/work/cuprum/cuprum/conftest.py'.
conftest.py:29: in <module>
    from tests.helpers.workflow import (
tests/helpers/workflow.py:7: in <module>
    import yaml
E   ModuleNotFoundError: No module named 'yaml'
make: *** [Makefile:211: test-markdown-format] Error 4
Error: Process completed with exit code 2.
~~~

https://github.com/leynos/cuprum/actions/runs/33998594606/job/101393326823?pr=325

Seek a systemic fix rather than tactical. Ask yourself, can this happen again or happen elsewhere? If so, think about a long term fix of the underlying issue.

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one exists already. (Treat o11y, code safety, documentation and validation coverage as in scope).

### ❌ Failed checks (3 errors)

| Check name | Status | Explanation | Resolution |
| :---: | :--- | :--- | :--- |
| Module-Level Documentation | ❌ Error | Update the module-level docstring in `cuprum/unittests/test_extension_ci_contract.py`. The pull request adds shared-actions reference validation and `lint-test` installer contract tests, but the uncha… | Extend the module docstring to document both contract areas: extension build ordering and shared-actions validation. State that the module parses `.github/workflows/ci.yml` and related workflows, verifies immutable shared-action targets, an… |
| Testing (Unit And Behavioural) | ❌ Error | The added tests cover the YAML contract, regex edge cases, ordering, toolchain inputs, and cache invariants. They do not exercise the changed workflow at its functional boundary. The pull request repl… | Add a black-box workflow integration test for the changed `lint-test` path. Run the workflow with a controlled event, assert a successful exit, and verify observable outputs such as the installed tool versions, available commands, and relev… |
| Unit Architecture | ❌ Error | The new workflow-reference query hides fallible file and YAML operations behind a pure-looking API. `_shared_action_references(workflow_name)` calls `workflow_document(workflow_name)`, which performs … | Separate workflow loading from reference extraction. Load and parse workflow documents at the test boundary, handle or clearly propagate `OSError` and YAML parsing errors there, and pass the parsed documents or source text into pure helpers… |
